src/Microsoft.DocAsCode.Dfm/RendererCreator.cs

```diff
@@ -0,0 +1,182 @@
+    using System.Collections.Immutable;
+    using System.Linq;
+    using System.Reflection;
+    using System.Reflection.Emit;
+    using System.Threading;
+
+    using Microsoft.DocAsCode.MarkdownLite;
+    using Microsoft.DocAsCode.Common;
+
+    public static class RendererCreator
+    {
+        #region Fields
+        private static readonly ModuleBuilder _module = CreateModule();
+        private static readonly MethodInfo MatchMethod = typeof(IDfmRendererPart).GetMethod(nameof(IDfmRendererPart.Match));
+        private static readonly MethodInfo RenderMethod = typeof(IDfmRendererPart).GetMethod(nameof(IDfmRendererPart.Render));
+        private static readonly ConstructorInfo BaseConstructor = typeof(PlugableRendererBase).GetConstructor(new[] { typeof(object) });
+        private static readonly MethodInfo BaseRenderMethod = typeof(PlugableRendererBase).GetMethod(nameof(PlugableRendererBase.BaseRender));
+        private static int _typeCounter;
+        #endregion
+
+        #region Public Methods
+
+        public static object CreateRenderer(object baseRenderer, IEnumerable<IDfmRendererPartProvider> partProviders, IReadOnlyDictionary<string, object> parameters)
+        {
+            if (baseRenderer == null)
+            {
+                throw new ArgumentNullException(nameof(baseRenderer));
+            }
+            if (partProviders == null)
+            {
+                return baseRenderer;
+            }
+            return CreateRendererNoCheck(baseRenderer, partProviders, parameters ?? ImmutableDictionary<string, object>.Empty);
+        }
+
+        #endregion
+
+        #region Private Methods
+
+        private static ModuleBuilder CreateModule()
+        {
+            const string name = "dfm-renderer";
+            var dynamicAssembly = AppDomain.CurrentDomain.DefineDynamicAssembly(
+                new AssemblyName(name),
+                AssemblyBuilderAccess.Run);
+            return dynamicAssembly.DefineDynamicModule(name);
+        }
+
+        private static object CreateRendererNoCheck(object baseRenderer, IEnumerable<IDfmRendererPartProvider> partProviders, IReadOnlyDictionary<string, object> parameters)
+        {
+            var partGroups = from provider in partProviders
+                             from part in provider.CreateParts(parameters)
+                             select new
+                             {
+                                 part,
+                                 types = Tuple.Create(part.MarkdownRendererType, part.MarkdownTokenType, part.MarkdownContextType)
+                             } into info
+                             group info.part by info.types;
+            if (!partGroups.Any())
+            {
+                return baseRenderer;
+            }
+            foreach (var item in from g in partGroups
+                                 where !ValidateKey(g.Key)
+                                 from item in g
+                                 select item)
```

Comment:
Seems the other style of linq is simple such as: partGroups.Where(p => !ValidateKey(p.Key)). So does the other one below.